source/common/compressor/compressor_impl.cc

```diff
@@ -0,0 +1,91 @@
+#include "common/compressor/compressor_impl.h"
+
+#include "envoy/common/exception.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Compressor {
+
+CompressorImpl::CompressorImpl()
+    : zstream_ptr_(new z_stream(), [](z_stream* z) {
+        deflateEnd(z);
+        delete z;
+      }) {
+  zstream_ptr_->zalloc = Z_NULL;
+  zstream_ptr_->zfree = Z_NULL;
+  zstream_ptr_->opaque = Z_NULL;
+}
+
+CompressorImpl::~CompressorImpl() {}
+
+void CompressorImpl::init(CompressionLevel comp_level, CompressionStrategy comp_strategy,
+                          int window_bits, uint8_t memory_level = 8) {
+  if (initialized_) {
+    throw EnvoyException("Compressor can only be initialized once");
+  }
+
+  const int res = deflateInit2(zstream_ptr_.get(), comp_level, Z_DEFLATED, window_bits,
+                               memory_level, comp_strategy);
+
+  if (res != Z_OK) {
+    // TODO: error mapping
```

Comment:
I think this is fine to leave as a TODO. Can you put your GH ID in the TODO? Format it like `TODO(gsagula)`

Reply:
I did it for the pending test case. Error mapping is quite simple in this case.